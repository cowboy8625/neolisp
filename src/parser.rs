use crate::ast::{Expr, Span, Spanned};
use crate::error::{Error, ErrorType};
use chumsky::prelude::*;

pub fn parse_expr() -> impl Parser<char, Spanned<Expr>, Error = Error> {
    recursive(|expr| {
        let comment = just(';').then(take_until(just('\n'))).padded();

        let int = text::int(10);

        let number = int
            .or(just('-').repeated().then(int).map(|(sign, num)| {
                let sign = if sign.len() % 2 == 0 { "" } else { "-" };
                format!("{sign}{num}")
            }))
            .map_with_span(|s: String, span: Span| {
                Spanned::from((Expr::Number(s.parse().unwrap()), span))
            })
            .padded();

        let boolean = just("true")
            .map_with_span(|_, span| Spanned::from((Expr::Bool(true), span)))
            .or(just("false").map_with_span(|_, span| (Expr::Bool(false), span).into()))
            .padded();

        let punctuation = one_of(r#"!$,_-./:;?+<=>#%&*@[\]{|}`^~"#);
        let letters = one_of("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ");
        let digit = one_of("0123456789");

        let symbol = choice((letters.clone(), punctuation.clone()))
            .then(choice((letters, punctuation, digit)).repeated())
            .padded()
            .map_with_span(|(start, end), span| {
                (
                    Expr::Symbol(format!("{start}{}", end.iter().collect::<String>())),
                    span,
                )
                    .into()
            });

        let string = just('"')
            .ignore_then(choice((
                none_of('"')
                    .repeated()
                    .at_least(1)
                    .collect::<String>()
                    .map(|s| s.replace("\\t", "\t"))
                    .map(|s| s.replace("\\n", "\n"))
                    .map_with_span(|s, span| (Expr::String(s), span).into())
                    .then_ignore(just('"')),
                just('"').map_with_span(|_, span| (Expr::String("".to_string()), span).into()),
            )))
            .padded();

        let atom = number
            .or(string)
            .or(boolean)
            .or(symbol)
            .or(expr.clone())
            .padded_by(comment.repeated());

        let quoted = just('\'')
            .ignore_then(atom.clone())
            .padded_by(comment.repeated())
            .map_with_span(|e: Spanned<Expr>, span: Span| {
                (
                    Expr::List(vec![
                        (Expr::Symbol("quote".to_string()), span.clone()).into(),
                        e,
                    ]),
                    span,
                )
                    .into()
            });

        let list = atom
            .repeated()
            .delimited_by(
                just('(').map_err(|e| Error::with_label(e, ErrorType::MissingOpeningParenthesis)),
                just(')').map_err(|e| Error::with_label(e, ErrorType::MissingClosingParenthesis)),
            )
            .padded_by(comment.repeated())
            .padded()
            .map_with_span(|list, span| Spanned::from((Expr::List(list), span)));

        list.or(quoted)
    })
}

pub fn parser() -> impl Parser<char, Vec<Spanned<Expr>>, Error = Error> {
    parse_expr().repeated().padded().then_ignore(end())
}
