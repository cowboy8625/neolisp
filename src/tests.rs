use crate::parser;
use chumsky::Parser;

macro_rules! snapshot_parse {
    ($name:tt, $contents:tt) => {
        #[test]
        fn $name() {
            let mut settings = insta::Settings::clone_current();
            settings.set_snapshot_path("testdata/output/");
            settings.bind(|| {
                let ast = format!("{:#?}", parser::parser().parse($contents));
                insta::assert_snapshot!(ast);
            });
        }
    };
}
snapshot_parse!(parse_test_ident1, "(hello-world)");
snapshot_parse!(parse_test_ident2, "(*ident*)");
snapshot_parse!(parse_test_ident3, "(<*ident/1234567890*>!?)");
snapshot_parse!(
    parse_test_bool_comment,
    "
; comment
(true)"
);
snapshot_parse!(parse_test_bool_true, "(true)");
snapshot_parse!(parse_test_bool_false, "(false)");
snapshot_parse!(parse_test_number, "(1)");
snapshot_parse!(parse_test_string, r#"("hello")"#);
snapshot_parse!(parse_test_string_empty, r#"("")"#);
snapshot_parse!(parse_test_string_ecape_char, r#"("\"")"#);
snapshot_parse!(parse_test_string_newline, r#"("hello\n")"#);
snapshot_parse!(parse_test_string_tab, r#"("hello\t")"#);
snapshot_parse!(parse_test_symbol_plus, "(+ 1 2)");
snapshot_parse!(parse_test_symbol_minus, "(+ 1 2)");
snapshot_parse!(parse_test_symbol_gt, "(> 1 2)");
snapshot_parse!(parse_test_symbol_lt, "(< 1 2)");
snapshot_parse!(parse_test_symbol_lte, "(<= 1 2)");
snapshot_parse!(parse_test_symbol_gte, "(>= 1 2)");
snapshot_parse!(parse_test_list, "((+ 1 2) (+ 1 2))");
snapshot_parse!(parse_test_lambda, "(lambda (x y) (+ x y))");
snapshot_parse!(parse_test_def_var, "(var x 1)");
snapshot_parse!(parse_test_def_func, "(fn add (x y) (+ x y))");
snapshot_parse!(parse_test_quote, "(var x '(1 2 3))");
snapshot_parse!(
    parse_test_program_with_many_single_list,
    " (x y) (+ x y) (x y z) "
);

macro_rules! snapshot_emit {
    ($name:tt, $contents:tt) => {
        #[test]
        fn $name() {
            let mut settings = insta::Settings::clone_current();
            settings.set_snapshot_path("testdata/output/");
            settings.bind(|| {
                let mut symbol_table = crate::symbol_table::SymbolTable::default();

                let Ok(Some(instructions)) = crate::compiler::Compiler::default()
                    .no_main(true)
                    .compile($contents, &mut symbol_table)
                else {
                    let msg = format!("Failed to compile: '{}'\n{}", stringify!($name), $contents);
                    return insta::assert_snapshot!(msg);
                };
                let string = crate::compiler::format_instructions(&instructions);

                insta::assert_snapshot!(string);
            });
        }
    };
    (enable_test $name:tt, $contents:tt) => {
        #[test]
        fn $name() {
            let mut settings = insta::Settings::clone_current();
            settings.set_snapshot_path("testdata/output/");
            settings.bind(|| {
                let mut symbol_table = crate::symbol_table::SymbolTable::default();

                let Ok(Some(instructions)) = crate::compiler::Compiler::default()
                    .no_main(true)
                    .with_test(true)
                    .compile($contents, &mut symbol_table)
                else {
                    let msg = format!("Failed to compile: '{}'\n{}", stringify!($name), $contents);
                    return insta::assert_snapshot!(msg);
                };
                let string = crate::compiler::format_instructions(&instructions);

                insta::assert_snapshot!(string);
            });
        }
    };
}

snapshot_emit!(emit_nil, "nil");
snapshot_emit!(emit_float, "1");
snapshot_emit!(emit_string, r#""hello world\n""#);
snapshot_emit!(emit_quote, "'(1 2 3)");
snapshot_emit!(emit_add_function, "(fn add (x y) (+ x y))");
snapshot_emit!(emit_simple_let, "(let (x 0) (+ 1 x))");
snapshot_emit!(
    emit_let,
    r#"
    (let (x 1)
    (let (y 2)
    (let (z 3)
    (print x y z "\n"))))"#
);
snapshot_emit!(emit_if_else, "(if true 'then 'else)");
snapshot_emit!(
    emit_apply_function,
    "(fn apply (f x) (f x))
(apply (lambda (x) (+ x 321)) 123)"
);
snapshot_emit!(
    emit_fib_function,
    "(fn fib (n a b)
  (if (= n 0)
      a
      (fib (- n 1) b (+ a b))))"
);
snapshot_emit!(emit_run_code, "(test test-testing (assert (= (+ 1 2) 3)))");
snapshot_emit!(enable_test emit_test_code, "(test test-testing (assert (= (+ 1 2) 3)))");
