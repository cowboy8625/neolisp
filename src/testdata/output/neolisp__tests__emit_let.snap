---
source: src/tests.rs
expression: string
---
000000      0  push       1
00000A     10  set-local  0:"x" 10..20
000019     25  push       2
000023     35  set-local  1:"y" 25..35
000032     50  push       3
00003C     60  set-local  2:"z" 40..50
00004B     75  push       "\n"
000052     82  get-local  2:"z" 61..63
000061     97  get-local  1:"y" 59..61
000070    112  get-local  0:"x" 57..59
00007F    127  push       <builtin 17:print:51..67>
000096    150  call       4
000098    152  halt
